### PR TITLE
[APP-898] Search Results

### PR DESCRIPTION
**Pull request by phillipsgoncalves**
**What does this PR do?**

**PLEASE WATCH THE APP-897 PR BEFORE THIS ONE**

This PR aims at implementing the search results, since the previous one was regarding the search suggestions. 
Note that just like in the last PR, I am not using the Aptoide theme yet, hence why some sizes and colors might be wrong. This is something we will fix afterwards.

**Database changed?**

   No

**Where should the reviewer start?**

- [ ] SearchView.kt

**How should this be manually tested?**

Open the app and perform a search. Confirm everything is ok.

**What are the relevant tickets?**

  Tickets related to this pull-request: [APP-898](https://aptoide.atlassian.net/browse/APP-898)

**Questions:**

   Does this add new dependencies which need to be added? (Eg. new keys, etc.) 

**Code Review Checklist**

- [ ] Documentation on public interfaces
- [ ] Database changed? If yes - Migration?
- [ ] Remove comments & unused code & forgotten testing Logs
- [ ] Codestyle
- [ ] New Kotlin code has unit tests
- [ ] New flows in presenters unit tests
- [ ] Mappers/Validators with any kind of logic unit tests
- [ ] Functional tests pass

### CODE REVIEW

File: feature_search/src/main/java/cm/aptoide/pt/feature_search/data/network/service/SearchRetrofitService.kt [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  override suspend fun getAutoCompleteSuggestions(keyword: String): Response<SearchAutoCompleteSuggestionsResponse> {
+    return autoCompleteSearchSuggestionsService.getAutoCompleteSuggestions(keyword)
+  }
+
+  override suspend fun searchApp(keyword: String): Response<BaseV7DataListResponse<SearchAppJsonList>> {
+    return searchAppRetrofitService.searchApp(keyword, 15)
+  }
+
+  interface AutoCompleteSearchRetrofitService {
+    @GET("/v1/suggestion/app/{query}")
```
Comment (jdandrade) [inline feature_search/src/main/java/cm/aptoide/pt/feature_search/data/network/service/SearchRetrofitService.kt:35]: v1 should probably be next to the baseUrl? what do you think? at least for the ws75.aptoide.com v7 we did that.

File: feature_search/src/main/java/cm/aptoide/pt/feature_search/di/RepositoryModule.kt [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    return AptoideSearchRepository(localSearchHistoryRepository, remoteSearchRepository)
+  }
+
+  @Singleton
+  @Provides
+  fun provideRemoteSearchRepository(
+    @RetrofitBuzz retrofitBuzz: Retrofit,
+    @RetrofitV7 retrofitV7: Retrofit
+  ): RemoteSearchRepository {
+    //return FakeRemoteSearchRepository()
```
Comment (jdandrade) [inline feature_search/src/main/java/cm/aptoide/pt/feature_search/di/RepositoryModule.kt:39]: Remove Comment, max we can do is to use the javadoc to let people know other implementations as fake.

File: feature_search/src/main/java/cm/aptoide/pt/feature_search/di/RepositoryModule.kt [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    return SearchRetrofitService(
+      retrofitBuzz.create(SearchRetrofitService.AutoCompleteSearchRetrofitService::class.java),
+      retrofitV7.create(SearchRetrofitService.SearchAppRetrofitService::class.java)
+    )
+  }
+
+  @Singleton
+  @Provides
+  fun provideLocalSearchHistoryRepository(database: SearchHistoryDatabase): LocalSearchHistoryRepository {
+    //return FakeLocalSearchHistory()
```
Comment (jdandrade) [inline feature_search/src/main/java/cm/aptoide/pt/feature_search/di/RepositoryModule.kt:49]: same

File: feature_search/src/main/java/cm/aptoide/pt/feature_search/domain/usecase/GetLocalTopDownloadedAppsUseCase.kt [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+  suspend fun getTopDownloadsRepository(): Result<List<LocalTopDownloadedApp>> {
+    return try {
+      //Result.Success(topDownloadsRepository.getLocalTopDownloadedApps())
+      Result.Success(Collections.emptyList())
+    } catch (e: Exception) {
+      Result.Error(e)
+    }
+  }
+}
```
Comment (jdandrade) [inline feature_search/src/main/java/cm/aptoide/pt/feature_search/domain/usecase/GetLocalTopDownloadedAppsUseCase.kt:21]: Comments, but I guess they will be removed when we go for implementation.

File: feature_search/src/main/res/layout/activity_main.xml [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+      android:layout_width="wrap_content"
+      android:layout_height="wrap_content"
+      android:text="Hello World!"
+      app:layout_constraintBottom_toBottomOf="parent"
+      app:layout_constraintLeft_toLeftOf="parent"
+      app:layout_constraintRight_toRightOf="parent"
+      app:layout_constraintTop_toTopOf="parent"
+      />
+
+</androidx.constraintlayout.widget.ConstraintLayout>
```
Comment (jdandrade) [inline feature_search/src/main/res/layout/activity_main.xml:19]: remove

File: feature_search/src/main/res/mipmap-anydpi-v26/ic_launcher_round.xml [hunk 6]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
+<adaptive-icon xmlns:android="http://schemas.android.com/apk/res/android">
+  <background android:drawable="@drawable/ic_launcher_background" />
+  <foreground android:drawable="@drawable/ic_launcher_foreground" />
+</adaptive-icon>
```
Comment (jdandrade) [inline feature_search/src/main/res/mipmap-anydpi-v26/ic_launcher_round.xml:5]: remove?

File: feature_search/src/main/res/mipmap-anydpi-v26/ic_launcher.xml [hunk 7]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
+<adaptive-icon xmlns:android="http://schemas.android.com/apk/res/android">
+  <background android:drawable="@drawable/ic_launcher_background" />
+  <foreground android:drawable="@drawable/ic_launcher_foreground" />
+</adaptive-icon>
```
Comment (jdandrade) [inline feature_search/src/main/res/mipmap-anydpi-v26/ic_launcher.xml:5]: remove

File: feature_search/src/main/res/values-night/themes.xml [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+    <item name="colorOnPrimary">@color/black</item>
+    <!-- Secondary brand color. -->
+    <item name="colorSecondary">@color/teal_200</item>
+    <item name="colorSecondaryVariant">@color/teal_200</item>
+    <item name="colorOnSecondary">@color/black</item>
+    <!-- Status bar color. -->
+    <item name="android:statusBarColor" tools:targetApi="l">?attr/colorPrimaryVariant</item>
+    <!-- Customize your theme here. -->
+  </style>
+</resources>
```
Comment (jdandrade) [inline feature_search/src/main/res/values-night/themes.xml:16]: removE?

File: feature_search/src/main/res/values/colors.xml [hunk 9]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+  <color name="purple_200">#FFBB86FC</color>
+  <color name="purple_500">#FF6200EE</color>
+  <color name="purple_700">#FF3700B3</color>
+  <color name="teal_200">#FF03DAC5</color>
+  <color name="teal_700">#FF018786</color>
+  <color name="black">#FF000000</color>
+  <color name="white">#FFFFFFFF</color>
+</resources>
```
Comment (jdandrade) [inline feature_search/src/main/res/values/colors.xml:10]: remove?

File: feature_search/src/main/res/values/themes.xml [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+    <item name="colorOnPrimary">@color/white</item>
+    <!-- Secondary brand color. -->
+    <item name="colorSecondary">@color/teal_200</item>
+    <item name="colorSecondaryVariant">@color/teal_700</item>
+    <item name="colorOnSecondary">@color/black</item>
+    <!-- Status bar color. -->
+    <item name="android:statusBarColor" tools:targetApi="l">?attr/colorPrimaryVariant</item>
+    <!-- Customize your theme here. -->
+  </style>
+</resources>
```
Comment (jdandrade) [inline feature_search/src/main/res/values/themes.xml:16]: remove

File: feature_search/src/test/java/cm/aptoide/pt/feature_search/ExampleUnitTest.kt [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+ * Example local unit test, which will execute on the development machine (host).
+ *
+ * See [testing documentation](http://d.android.com/tools/testing).
+ */
+class ExampleUnitTest {
+  @Test
+  fun addition_isCorrect() {
+    assertEquals(4, 2 + 2)
+  }
+}
```
Comment (jdandrade) [inline feature_search/src/test/java/cm/aptoide/pt/feature_search/ExampleUnitTest.kt:17]: remove

**Review by jdandrade — CHANGES_REQUESTED** [review]
Just not sure about one thing but I think that we can go with this first an refactor it later if necessary: We separated TopSearches from SearchSuggestions. I think that everything we list there are suggestions, just that depending on the query (non-existent, existent) they are different sources, this should all be under the same manager. We should discuss this.

File: feature_search/src/main/java/cm/aptoide/pt/feature_search/presentation/search/SearchViewModel.kt [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+  val searchSuggestions: SearchSuggestions,
+  val searchTextInput: String = "",
+  val isLoading: Boolean = false,
+  val hasErrors: Boolean = false,
+  val searchAppBarState: SearchAppBarState = SearchAppBarState.CLOSED,
+  val searchResults: List<SearchApp> = emptyList()
+) {
+
+  fun toUiState(): SearchUiState =
+    //if (!hasErrors) {
```
Comment (jdandrade) [inline feature_search/src/main/java/cm/aptoide/pt/feature_search/presentation/search/SearchViewModel.kt:128]: comments

**Review by jdandrade — CHANGES_REQUESTED** [review]
Strings are hardcoded, we need to review this with localisation team after, probably most strings are already in dev but they have to be merged to dev-v10 and then here.

File: feature_search/src/main/java/cm/aptoide/pt/feature_search/data/AptoideSearchRepository.kt [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+    return flow {
+      val autoCompleteResponse = remoteSearchRepository.getAutoCompleteSuggestions(keyword)
+      if (autoCompleteResponse.isSuccessful) {
+        autoCompleteResponse.body()?.data?.let {
+          emit(AutoCompleteResult.Success(it.map { suggestion -> AutoCompletedApp(suggestion) }))
+        }
+      } else {
+        emit(AutoCompleteResult.Error(IllegalStateException()))
+      }
+    }.flowOn(Dispatchers.IO)
```
[the remark this answers is not in the document]
Comment (phillipsgoncalves) [reply feature_search/src/main/java/cm/aptoide/pt/feature_search/data/AptoideSearchRepository.kt:72]: We discussed that we should remove the catch from the repository and all exceptions should be caught on the viewmodel in order to be able to update the view.

File: feature_search/src/main/java/cm/aptoide/pt/feature_search/data/network/service/SearchRetrofitService.kt (hunk 1 above)
Comment (phillipsgoncalves) [reply feature_search/src/main/java/cm/aptoide/pt/feature_search/data/network/service/SearchRetrofitService.kt:35]: Discussion: We decided we should have the version here, to make it easier to manage versions.

File: feature_search/src/main/java/cm/aptoide/pt/feature_search/presentation/search/SearchView.kt [hunk 14]
```diff
@@ -0,0 +1,411 @@
...
+        overflow = TextOverflow.Ellipsis
+      )
+      RatingSearchView(searchApp.rating)
+      Text(
+        text = searchApp.downloads.toString() + " downloads",
+        maxLines = 1,
+        fontSize = MaterialTheme.typography.overline.fontSize
+      )
+    }
+    if (searchApp.malware == "TRUSTED") {
```
[the remark this answers is not in the document]
Comment (phillipsgoncalves) [reply feature_search/src/main/java/cm/aptoide/pt/feature_search/presentation/search/SearchView.kt:141]: Discussion: talk with Silvana when she comes back

**Pull request merged**